### PR TITLE
rgw_sal_motr: [CORTX-31439] Add support for copy object in same zonegroup

### PR DESCRIPTION
**Pull request by SeagateChaDeepak**
## Problem :
No support for copy object in current implementation.

## Solution :
1. Use ReadOp class to read from source object.
2. Use Atomic writer class to write to destination object.

Signed-off-by: Deepak Mahale <deepak.mahale@seagate.com>

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/rgw/rgw_sal_motr.cc [hunk 1]
```diff
@@ -1885,6 +1884,121 @@ int MotrObject::delete_obj_aio(const DoutPrefixProvider* dpp, RGWObjState* astat
   return 0;
 }
 
+int MotrCopyObj_CB::handle_data(bufferlist& bl, off_t bl_ofs, off_t bl_len)
+{
+  int rc = 0;
+  ldpp_dout(m_dpp, 20) << "Offset = " << bl_ofs << " Length = " << bl_len << dendl;
+  
+  rc = m_dst_writer->process(std::move(bl), bl_ofs);
```
Comment (sjain09) [inline src/rgw/rgw_sal_motr.cc:1892]: we should not pass bl_ofs here, this is local offset of bl.
in `MotrAtomicWriter::process(bufferlist&& data, uint64_t offset)` , `offset` is offset of the entire object and in `data`, data is always present from 0th index, which is not the case with `bl`
We need to create a local `data`, copy `bl` starting from bl_ofs to bl_len.
You can refer `RGWRESTStreamOutCB::handle_data(bufferlist& bl, off_t bl_ofs, off_t bl_len)` function where this is handled.

Update the `this->offfset `
`this->offfset += bl_len;`

Then pass this->offfset in the process function.

File: src/rgw/rgw_sal_motr.h [hunk 2]
```diff
@@ -501,6 +501,27 @@ struct AccumulateIOCtxt{
...
+public:
+  virtual int handle_data(bufferlist& bl, off_t bl_ofs, off_t bl_len) override { return 0; }
+  MotrCopyObj_Filter() {}
+  virtual ~MotrCopyObj_Filter() override {}
+};
+
+class MotrCopyObj_CB : public MotrCopyObj_Filter
+{
+  const DoutPrefixProvider* m_dpp;
+  std::shared_ptr<rgw::sal::Writer> m_dst_writer;
```
Comment (sjain09) [inline src/rgw/rgw_sal_motr.h:514]: Add one more variable as `uint64 offset` to track the data written to object so far.

File: src/rgw/rgw_sal_motr.cc [hunk 3]
```diff
@@ -1885,6 +1884,121 @@ int MotrObject::delete_obj_aio(const DoutPrefixProvider* dpp, RGWObjState* astat
...
+  ldpp_dout(dpp, 20) << "Dest Object Name : " << dest_object->get_key().get_oid() << dendl;
+  std::unique_ptr<rgw::sal::Object::ReadOp> read_op = this->get_read_op(&obj_ctx);
+
+  std::shared_ptr<rgw::sal::Writer> dst_writer = store->get_atomic_writer(dpp, y,
+          std::unique_ptr<rgw::sal::Object>(dest_object),
+          user->get_id(), obj_ctx,
+          &dest_placement, olh_epoch, *version_id);
+
+  // prepare write op
+  rc = dst_writer->prepare(y);
```
Comment (sjain09) [inline src/rgw/rgw_sal_motr.cc:1942]: Do the read prepare first before creating the write object.

File: src/rgw/rgw_sal_motr.cc (hunk 1 above)
Comment (SeagateChaDeepak) [reply src/rgw/rgw_sal_motr.cc]: made some changes @sjain09. Please review.

File: src/rgw/rgw_sal_motr.h (hunk 2 above)
Comment (SeagateChaDeepak) [reply src/rgw/rgw_sal_motr.h:514]: Made some changes @sjain09. Please review

File: src/rgw/rgw_sal_motr.cc (hunk 3 above)
Comment (SeagateChaDeepak) [reply src/rgw/rgw_sal_motr.cc:1941]: Does the sequence matter? or is it just for readability?

**Review by sjain09 — CHANGES_REQUESTED** [review]

File: src/rgw/rgw_sal_motr.cc [hunk 4]
```diff
@@ -1885,6 +1868,136 @@ int MotrObject::delete_obj_aio(const DoutPrefixProvider* dpp, RGWObjState* astat
   return 0;
 }
 
+int MotrCopyObj_CB::handle_data(bufferlist& bl, off_t bl_ofs, off_t bl_len)
+{
+  int rc = 0;
+  ldpp_dout(m_dpp, 20) << "Offset = " << bl_ofs << " Length = " << bl_len << dendl;
```
[the remark this answers is not in the document]
Comment (SeagateChaDeepak) [reply src/rgw/rgw_sal_motr.cc:1874]: Done

File: src/rgw/rgw_sal_motr.cc [hunk 5]
```diff
@@ -1885,6 +1868,136 @@ int MotrObject::delete_obj_aio(const DoutPrefixProvider* dpp, RGWObjState* astat
...
+
+  //offset is zero and bufferlength is equal to bl_len
+  if (!bl_ofs && bl_len == bl.length()) {
+    rc = m_dst_writer->process(std::move(bl), write_offset);
+    if(rc < 0){
+      ldpp_dout(m_dpp, 20) << "ERROR: writer process bl_ofs=0 && " <<
+                          "bl_len = bl.length failed rc = " << rc << dendl;
+    }
+    write_offset += (bl_len - bl_ofs);
```
[the remark this answers is not in the document]
Comment (SeagateChaDeepak) [reply src/rgw/rgw_sal_motr.cc:1883]: Done

File: src/rgw/rgw_sal_motr.cc [hunk 6]
```diff
@@ -1885,6 +1868,136 @@ int MotrObject::delete_obj_aio(const DoutPrefixProvider* dpp, RGWObjState* astat
...
+int MotrCopyObj_CB::handle_data(bufferlist& bl, off_t bl_ofs, off_t bl_len)
+{
+  int rc = 0;
+  ldpp_dout(m_dpp, 20) << "Offset = " << bl_ofs << " Length = " << bl_len << dendl;
+
+  //offset is zero and bufferlength is equal to bl_len
+  if (!bl_ofs && bl_len == bl.length()) {
+    rc = m_dst_writer->process(std::move(bl), write_offset);
+    if(rc < 0){
+      ldpp_dout(m_dpp, 20) << "ERROR: writer process bl_ofs=0 && " <<
```
[the remark this answers is not in the document]
Comment (SeagateChaDeepak) [reply src/rgw/rgw_sal_motr.cc:1880]: Done

File: src/rgw/rgw_sal_motr.cc [hunk 7]
```diff
@@ -1885,6 +1868,136 @@ int MotrObject::delete_obj_aio(const DoutPrefixProvider* dpp, RGWObjState* astat
...
+{
+  int rc = 0;
+  ldpp_dout(dpp, 20) << "Src Object Name : " << this->get_key().get_oid() << dendl;
+  ldpp_dout(dpp, 20) << "Dest Object Name : " << dest_object->get_key().get_oid() << dendl;
+  std::unique_ptr<rgw::sal::Object::ReadOp> read_op = this->get_read_op(&obj_ctx);
+
+  std::shared_ptr<rgw::sal::Writer> dst_writer = store->get_atomic_writer(dpp, y,
+          std::unique_ptr<rgw::sal::Object>(dest_object),
+          user->get_id(), obj_ctx,
+          &dest_placement, olh_epoch, *version_id);
```
[the remark this answers is not in the document]
Comment (SeagateChaDeepak) [reply src/rgw/rgw_sal_motr.cc]: Done

File: src/rgw/rgw_sal_motr.cc [hunk 8]
```diff
@@ -1885,6 +1868,174 @@ int MotrObject::delete_obj_aio(const DoutPrefixProvider* dpp, RGWObjState* astat
...
+  }
+
+  real_time time = ceph::real_clock::now();
+  if(mtime){
+    *mtime = time;
+  }
+
+  //fetch etag.
+  rgw_bucket_dir_entry ent;
+  rc = this->get_bucket_dir_ent(dpp, ent);
```
[the remark this answers is not in the document]
Comment (sjain09) [reply src/rgw/rgw_sal_motr.cc:2005]: You can add this code to read the etag from read_op:
```
int MotrObject::MotrReadOp::get_attr(const DoutPrefixProvider* dpp, const char* name, bufferlist& dest, optional_yield y)
{
  if (source == nullptr)
    return -ENODATA;
  rgw::sal::Attrs attrs = source->get_attrs()
  auto iter = attrs.find(name);
  if (iter != attrs.end()) {
    dest = iter->second;
    return 0;
  }
  return -ENODATA;
}

```

and call `read_op->get_attr(RGW_ATTR_ETAG)` to fetch etag

**Review by sjain09 — CHANGES_REQUESTED** [review]

File: src/rgw/rgw_sal_motr.cc (hunk 3 above)
Comment (SeagateChaDeepak) [reply src/rgw/rgw_sal_motr.cc]: Done

File: src/rgw/rgw_sal_motr.cc [hunk 9]
```diff
@@ -1885,6 +1868,174 @@ int MotrObject::delete_obj_aio(const DoutPrefixProvider* dpp, RGWObjState* astat
...
+  if (rc < 0){
+    ldpp_dout(dpp, 20) << "ERROR: get_bucket_dir_ent failed rc = " << rc << dendl;
+    return rc;
+  }
+
+  std::string etag_str;
```
[the remark this answers is not in the document]
Comment (SeagateChaDeepak) [reply src/rgw/rgw_sal_motr.cc:2011]: In case etag is passed as null??

[the remark this answers is not in the document]
Comment (sjain09) [reply src/rgw/rgw_sal_motr.cc:2011]: This is a output parameter and we are fetching etag from read_op. so you can directly pass that into the function and assign the same to the output etag.

File: src/rgw/rgw_sal_motr.cc (hunk 8 above)
[the remark this answers is not in the document]
Comment (SeagateChaDeepak) [reply src/rgw/rgw_sal_motr.cc:2005]: Addressed in a different way

File: src/rgw/rgw_sal_motr.cc [hunk 10]
```diff
@@ -1914,7 +2065,67 @@ int MotrObject::copy_object(RGWObjectCtx& obj_ctx,
     const DoutPrefixProvider* dpp,
     optional_yield y)
 {
-      return 0;
+  int rc = 0;
+  auto& src_zonegrp = src_bucket->get_info().zonegroup;
+  auto& dest_zonegrp = dest_bucket->get_info().zonegroup;
+
+  if(src_zonegrp.compare(dest_zonegrp) == 0){
```
[the remark this answers is not in the document]
Comment (SeagateChaDeepak) [reply src/rgw/rgw_sal_motr.cc:2072]: Done

File: src/rgw/rgw_sal_motr.cc (hunk 8 above)
[the remark this answers is not in the document]
Comment (sjain09) [reply src/rgw/rgw_sal_motr.cc:2286]: you can do string assignment like this : 
```
bufferlist bl;
read_op->get_attr(dpp, RGW_ATTR_ETAG, bl, y)
string etag_str;
etag_str.assign(bl.c_str(), bl.length());
```

File: src/rgw/rgw_sal_motr.cc (hunk 9 above)
[the remark this answers is not in the document]
Comment (SeagateChaDeepak) [reply src/rgw/rgw_sal_motr.cc]: Done

**Review by sjain09 — APPROVED** [review]

File: src/rgw/rgw_sal_motr.cc [hunk 11]
```diff
@@ -1899,7 +1882,14 @@ int MotrObject::MotrReadOp::iterate(const DoutPrefixProvider* dpp, int64_t off,
 
 int MotrObject::MotrReadOp::get_attr(const DoutPrefixProvider* dpp, const char* name, bufferlist& dest, optional_yield y)
 {
-  //return 0;
+  if (source == nullptr)
```
Comment (sachinpunadikar) [inline src/rgw/rgw_sal_motr.cc:1885]: This code can be as below:
  if (source != nullptr) {
     rgw::sal::Attrs &attrs = source->get_attrs();
     auto iter = attrs.find(name);
     if (iter != attrs.end()) {
         dest = iter->second;
         return 0;
      }
  }
  return -ENODATA;

**Comment by sachinpunadikar** [comment]
curious to know where dest_object is getting creating ? was not able to locate where we are calling create_mobj during execution of copy_object
@sjain09 , @SeagateChaDeepak

**Comment by sjain09** [comment]
> curious to know where dest_object is getting creating ? was not able to locate where we are calling create_mobj during execution of copy_object @sjain09 , @SeagateChaDeepak

@sachinpunadikar , we making use of existing `MotrAtomicWriter` class to create and write motr object. The `process()` function internally calls `create_mobj()` and write object data in motr. Finally we call `complete()` function that will write the metadata into the index.

**Review by sachinpunadikar — APPROVED** [review]
LGTM

File: src/rgw/rgw_sal_motr.cc [hunk 12]
```diff
@@ -1480,26 +1480,10 @@ int MotrObject::get_obj_state(const DoutPrefixProvider* dpp, RGWObjectCtx* rctx,
   *_state = state;
 
   // Get object's metadata (those stored in rgw_bucket_dir_entry).
-  bufferlist bl;
```
Comment (DPG17) [inline src/rgw/rgw_sal_motr.cc:1483]: @SeagateChaDeepak : Why is this cache code removed?

**Review by DPG17 — APPROVED** [review]

File: src/rgw/rgw_sal_motr.cc (hunk 11 above)
Comment (SeagateChaDeepak) [reply src/rgw/rgw_sal_motr.cc:1885]: Hi Sachin,
this will be addressed in the future PR.

File: src/rgw/rgw_sal_motr.cc (hunk 12 above)
Comment (sjain09) [reply src/rgw/rgw_sal_motr.cc:1486]: As get_bucket_dir_ent() function handles all the version related changes, this should be fine.

**Pull request merged**